### ISSUE TITLE
General Help (was: Unable to import `undetected_chromedriver`: `collections` has no `Mapping`)

### ISSUE THREAD

**Issue by chrishat34** [2e0byo/YADC#1]
I get the following errors when fireing this up, any advice would be most welcome:
(venv) C:\Development from 16-11-2021\YADC>python -m yadc.main
Traceback (most recent call last):
  File "C:\Development from 16-11-2021\python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Development from 16-11-2021\python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Development from 16-11-2021\YADC\yadc\main.py", line 6, in <module>
    from .browser import Browser, TorBrowser
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 10, in <module>
    import undetected_chromedriver as uc
  File "C:\Development from 16-11-2021\YADC\venv\lib\site-packages\undetected_chromedriver\__init__.py", line 31, in <module>
    from . import v2
  File "C:\Development from 16-11-2021\YADC\venv\lib\site-packages\undetected_chromedriver\v2.py", line 24, in <module>
    from .cdp import CDP
  File "C:\Development from 16-11-2021\YADC\venv\lib\site-packages\undetected_chromedriver\cdp.py", line 6, in <module>
    from collections import Mapping, Sequence
ImportError: cannot import name 'Mapping' from 'collections' (C:\Development from 16-11-2021\python310\lib\collections\__init__.py)

**Comment by 2e0byo** [comment]
This is a bug in undetected_chromedriver:

```
 DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
```

i.e. python 3.10 has moved `Mapping` to a special module `collections.abc` for Abstract Base Classes (i.e. classes which other things conceptually 'inherit from', although they don't in reality).  You have a few options:

1. run `poetry update undetected-chromedriver` in case it's been fixed upstream, and then try again.
2. file a bug against `undetected-chromedriver`
3. install and use python 3.9

Do try 1 & 2, but 3. is probably easiest.

Incidentally, I'm just about to push up a more robust `kill()` implementation as I've discovered some systems don't kill the browser process when the containing shell is killed.

**Comment by 2e0byo** [comment]
This is already reported and there's a [pr](https://github.com/ultrafunkamsterdam/undetected-chromedriver/pull/294)

**Comment by 2e0byo** [comment]
Since we do not actually use `undetected_chromedriver` and it was nontrivial to get it running, I have simply removed the dep.  Thus if you pull this should go away.

**Comment by chrishat34** [comment]
I still cannot get it going, it is possible you could help me directly. Of course i would pay an appropriate fee for your time. If possible please pm me an email to jchrishoward910@gmail.com.

**Comment by 2e0byo** [comment]
@chrishat34  I'm not doing that, but if you post the error msg I'll certainly have a look at it. I've not tried on anything except Linux, mind.  Firstly, be sure you're using the latest, i.e. pull again (and be sure it merges if you've updated anything locally).  Then post the error traceback here and I'll have a look :)

**Comment by chrishat34** [comment]
Okky docky.
i have got furthr than i have ever got so that good, but....
yet installed buster as it suggested it was optional, but i get this error so far.
(venv) C:\Development from 16-11-2021\YADC>python -m yadc.main
Traceback (most recent call last):
  File "C:\Development from 16-11-2021\python399\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Development from 16-11-2021\python399\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Development from 16-11-2021\YADC\yadc\main.py", line 32, in <module>
    TorBrowser(buster=Path("buster"), chrome="chromium"), [driver], notify=notify
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 171, in __init__
    raise BrowserError("Please unzip buster and pass the dir.")
yadc.browser.BrowserError: Please unzip buster and pass the dir.

**Comment by chrishat34** [comment]
...thank you

**Comment by chrishat34** [comment]
Hi, made some good progress, now get too....
(venv) C:\Development from 16-11-2021\YADC>python -m yadc.main
2021-11-19 18:33:10 DESKTOP-4V4OMAL Browser-1[2256] INFO Starting search loop
2021-11-19 18:33:10 DESKTOP-4V4OMAL Browser-1[2256] INFO Starting Tor
2021-11-19 18:33:10 DESKTOP-4V4OMAL Browser-1[2256] ERROR [WinError 2] The system cannot find the file specified
Traceback (most recent call last):
  File "C:\Development from 16-11-2021\YADC\yadc\scraper.py", line 247, in __call__
    with self._browser as browser:
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 253, in __enter__
    self.launch_chrome()
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 328, in launch_chrome
    self.start_tor()
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 323, in start_tor
    proc = Popen(cmd, stdout=PIPE, encoding="utf8")
  File "C:\Development from 16-11-2021\python399\lib\subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Development from 16-11-2021\python399\lib\subprocess.py", line 1420, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified
/

**Comment by 2e0byo** [comment]
> installed buster as it suggested it was optional, but i get this error so far... `yadc.browser.BrowserError: Please unzip buster and pass the dir.`

Yes, buster is optional, but if you want captcha solution you have to unzip the downloaded package and tell chrome where to find it.  You clearly figured this out, but I guess the docs could do with some work.

(Incidentally using three backticks will get you code block (use three backticks on a line by themselves before and after, and add the language afterwards, here python, to get syntax highlighting.  Same goes for Stack Overflow.)

Your current issue is that you are trying to start `tor` without `tor` being in your path.  I don't have windows on hand to test, but assuming you have installed `tor`, find the path to something like `tor.exe`.  I have just modified the class definition, so you can pass it in to the constructor in your `main.py`:

```python
TorBrowser(..., tor="C:\Program Files\tor.exe")
```

And it should hopefully work for you.  You're pretty close.  You will likely need to tell it where chrome and chromedriver are, too. I wrote this script for my own use and don't have windows on hand, so sorry the documentation isn't better. But you're helping to make it better, so all is good :)

Alternatively you can replace `TorBrowser` with `Browser`, but you lose the advantage of `tor` scrambling your connections and you will likely get blocked after a while.  Worth trying just to test it though if tor won't work.  On my end tor setup was pretty painless.

**Comment by 2e0byo** [comment]
reopened again although this is off topic for this title and this is more of a general help issue.  Will close when it's running for you.

**Comment by chrishat34** [comment]
Thank you for all this help and sorry for being so useless.
I have the new browser.py file downloaded but cannot see where i place the path to my tor.exe in the code. My path is C:\Development from 16-11-2021\Tor Browser\Browser\TorBrowser\Tor - does it go in the .TorBrowser class def or ?
Probably ditto the paths chrome and chromedriver

**Comment by chrishat34** [comment]
ok, getting there..
```python
(venv) C:\Development from 16-11-2021\YADC>python -m yadc.main
2021-11-20 08:15:58 DESKTOP-4V4OMAL Browser-1[2604] INFO Starting search loop
2021-11-20 08:15:58 DESKTOP-4V4OMAL Browser-1[2604] INFO Starting Tor
2021-11-20 08:15:58 DESKTOP-4V4OMAL Browser-1[2604] INFO Started Tor
2021-11-20 08:15:58 DESKTOP-4V4OMAL Browser-1[2604] INFO Starting chrome
2021-11-20 08:15:58 DESKTOP-4V4OMAL Browser-1[2604] INFO This time we are Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/27.0.1453.93 Safari/537.36
2021-11-20 08:15:58 DESKTOP-4V4OMAL Browser-1[2604] INFO Waiting 10s for page to load and js to run.
'chromium' is not recognized as an internal or external command,
operable program or batch file.
2021-11-20 08:16:08 DESKTOP-4V4OMAL Browser-1[2604] ERROR Message: 'chromedriver' executable needs to be in PATH. Please see https://chromedriver.chromium.org/home
```

**Comment by 2e0byo** [comment]
you just need the path to the chromedriver executable to be in the same place in your `main.py`:

```python
s = Scraper(
                    TorBrowser(
                                        buster=Path("path/to/buster/dir"), 
                                        chrome="path/to/chrome",
                                        chromedriver="path/to/chromedriver",
                                        tor="path/to/tor",
                       ), 
                     [driver], 
                     notify=notify
)
```

I guess I could have made that a lot clearer!  When you get it working I'll update the docs.

Another thing you *may* have problems with because I have not tried on windows is killing the browser.  psutil claims to be cross-platform and handle it itself: lt me know if it works.

**Comment by chrishat34** [comment]
ok cool, progress.
here is my code block, ok?
```python
s = Scraper(
                    TorBrowser(
                                        buster=Path("C:\\Development from 16-11-2021\\YADC\\buster"),
                                        chrome=Path("C:\\Program Files\\Google\\Chrome\\Application"),
                                        chromedriver=Path("C:\\Development from 16-11-2021"),
                                        tor=Path("C:\\Development from 16-11-2021\\Tor Browser\\Browser\\TorBrowser\\Tor\\tor.exe"),
                       ),
                     [driver],
                     notify=notify
)
```
and the output is:
```python
(venv) C:\Development from 16-11-2021\YADC>python -m yadc.main
2021-11-20 11:21:07 DESKTOP-4V4OMAL Browser-1[10372] INFO Starting search loop
2021-11-20 11:21:07 DESKTOP-4V4OMAL Browser-1[10372] INFO Starting Tor
2021-11-20 11:21:07 DESKTOP-4V4OMAL Browser-1[10372] INFO Started Tor
2021-11-20 11:21:07 DESKTOP-4V4OMAL Browser-1[10372] INFO Starting chrome
2021-11-20 11:21:07 DESKTOP-4V4OMAL Browser-1[10372] INFO This time we are Mozilla/5.0 (Windows NT 5.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/27.0.1453.93 Safari/537.36
2021-11-20 11:21:07 DESKTOP-4V4OMAL Browser-1[10372] ERROR sequence item 0: expected str instance, WindowsPath found
Traceback (most recent call last):
  File "C:\Development from 16-11-2021\YADC\yadc\scraper.py", line 247, in __call__
    with self._browser as browser:
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 245, in __enter__
    self.launch_chrome()
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 331, in launch_chrome
    super().launch_chrome(extra_args=tor_args)
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 242, in launch_chrome
    self._proc = Popen(" ".join(cmd), shell=True)
TypeError: sequence item 0: expected str instance, WindowsPath found
Traceback (most recent call last):
  File "C:\Development from 16-11-2021\YADC\yadc\scraper.py", line 247, in __call__
    with self._browser as browser:
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 245, in __enter__
    self.launch_chrome()
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 331, in launch_chrome
    super().launch_chrome(extra_args=tor_args)
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 242, in launch_chrome
    self._proc = Popen(" ".join(cmd), shell=True)
TypeError: sequence item 0: expected str instance, WindowsPath found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Development from 16-11-2021\python399\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Development from 16-11-2021\python399\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Development from 16-11-2021\YADC\yadc\main.py", line 45, in <module>
    s()
  File "C:\Development from 16-11-2021\YADC\yadc\scraper.py", line 255, in __call__
    randsleep(30) # try not to be too predictable
  File "C:\Development from 16-11-2021\YADC\yadc\humanlike.py", line 27, in randsleep
    sleep(duration)
KeyboardInterrupt
^C
```

**Comment by 2e0byo** [comment]
Tor should be a string: no need for the Path.

Le sam. 20 nov. 2021 à 11:26, chrishat34 ***@***.***> a
écrit :

**Comment by 2e0byo** [comment]
Same for chrome and chrome driver.  See the type hints I browser.py. :)

**Comment by 2e0byo** [comment]
Right, I have just pushed a commit (untested) which should let you pass either paths or strs for any of the path args.  Typehints updated accordingly.

**Comment by chrishat34** [comment]
ok, here are, we are making progress, sorry its long winded...i changed the localhost address and didnt help

```python
(venv) C:\Development from 16-11-2021\YADC>python -m yadc.main
2021-11-20 13:43:41 DESKTOP-4V4OMAL Browser-1[8092] INFO Starting search loop
2021-11-20 13:43:41 DESKTOP-4V4OMAL Browser-1[8092] INFO Starting Tor
2021-11-20 13:43:41 DESKTOP-4V4OMAL Browser-1[8092] INFO Started Tor
2021-11-20 13:43:41 DESKTOP-4V4OMAL Browser-1[8092] INFO Starting chrome
2021-11-20 13:43:41 DESKTOP-4V4OMAL Browser-1[8092] INFO This time we are Mozilla/5.0 (Windows NT 6.4; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2225.0 Safari/537.36
2021-11-20 13:43:41 DESKTOP-4V4OMAL Browser-1[8092] INFO Waiting 10s for page to load and js to run.
'C:\Program' is not recognized as an internal or external command,
operable program or batch file.
2021-11-20 13:44:54 DESKTOP-4V4OMAL Browser-1[8092] ERROR Message: unknown error: cannot connect to chrome at localhost:8898
from chrome not reachable
Stacktrace:
Backtrace:
        Ordinal0 [0x007E06F3+2492147]
        Ordinal0 [0x00779BD1+2071505]
        Ordinal0 [0x00682340+1057600]
        Ordinal0 [0x0067739D+1012637]
        Ordinal0 [0x0069FF64+1179492]
        Ordinal0 [0x00699DCB+1154507]
        Ordinal0 [0x00699BB1+1153969]
        Ordinal0 [0x006C9DFF+1351167]
        Ordinal0 [0x006C9A6A+1350250]
        Ordinal0 [0x006C568B+1332875]
        Ordinal0 [0x006A21D4+1188308]
        Ordinal0 [0x006A302F+1191983]
        GetHandleVerifier [0x009667A6+1545030]
        GetHandleVerifier [0x00A1105C+2243580]
        GetHandleVerifier [0x0086BC97+518199]
        GetHandleVerifier [0x0086AD80+514336]
        Ordinal0 [0x0077ED2D+2092333]
        Ordinal0 [0x00782EE8+2109160]
        Ordinal0 [0x00783022+2109474]
        Ordinal0 [0x0078CB71+2149233]
        BaseThreadInitThunk [0x7582FA29+25]
        RtlGetAppContainerNamedObjectPath [0x779B7A9E+286]
        RtlGetAppContainerNamedObjectPath [0x779B7A6E+238]

Traceback (most recent call last):
  File "C:\Development from 16-11-2021\YADC\yadc\scraper.py", line 247, in __call__
    with self._browser as browser:
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 271, in __enter__
    driver = CaptchaChrome(
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 35, in __init__
    super().__init__(*args, **kwargs)
  File "C:\Development from 16-11-2021\venv\lib\site-packages\selenium\webdriver\chrome\webdriver.py", line 69, in __init__
    super(WebDriver, self).__init__(DesiredCapabilities.CHROME['browserName'], "goog",
  File "C:\Development from 16-11-2021\venv\lib\site-packages\selenium\webdriver\chromium\webdriver.py", line 93, in __init__
    RemoteWebDriver.__init__(
  File "C:\Development from 16-11-2021\venv\lib\site-packages\selenium\webdriver\remote\webdriver.py", line 266, in __init__
    self.start_session(capabilities, browser_profile)
  File "C:\Development from 16-11-2021\venv\lib\site-packages\selenium\webdriver\remote\webdriver.py", line 357, in start_session
    response = self.execute(Command.NEW_SESSION, parameters)
  File "C:\Development from 16-11-2021\venv\lib\site-packages\selenium\webdriver\remote\webdriver.py", line 418, in execute
    self.error_handler.check_response(response)
  File "C:\Development from 16-11-2021\venv\lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 243, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: unknown error: cannot connect to chrome at localhost:8898
from chrome not reachable
Stacktrace:
Backtrace:
        Ordinal0 [0x007E06F3+2492147]
        Ordinal0 [0x00779BD1+2071505]
        Ordinal0 [0x00682340+1057600]
        Ordinal0 [0x0067739D+1012637]
        Ordinal0 [0x0069FF64+1179492]
        Ordinal0 [0x00699DCB+1154507]
        Ordinal0 [0x00699BB1+1153969]
        Ordinal0 [0x006C9DFF+1351167]
        Ordinal0 [0x006C9A6A+1350250]
        Ordinal0 [0x006C568B+1332875]
        Ordinal0 [0x006A21D4+1188308]
        Ordinal0 [0x006A302F+1191983]
        GetHandleVerifier [0x009667A6+1545030]
        GetHandleVerifier [0x00A1105C+2243580]
        GetHandleVerifier [0x0086BC97+518199]
        GetHandleVerifier [0x0086AD80+514336]
        Ordinal0 [0x0077ED2D+2092333]
        Ordinal0 [0x00782EE8+2109160]
        Ordinal0 [0x00783022+2109474]
        Ordinal0 [0x0078CB71+2149233]
        BaseThreadInitThunk [0x7582FA29+25]
        RtlGetAppContainerNamedObjectPath [0x779B7A9E+286]
        RtlGetAppContainerNamedObjectPath [0x779B7A6E+238]

2021-11-20 13:45:25 DESKTOP-4V4OMAL Browser-1[8092] INFO Starting search loop
2021-11-20 13:45:25 DESKTOP-4V4OMAL Browser-1[8092] INFO Starting Tor
2021-11-20 13:45:25 DESKTOP-4V4OMAL Browser-1[8092] INFO Started Tor
2021-11-20 13:45:25 DESKTOP-4V4OMAL Browser-1[8092] INFO Starting chrome
2021-11-20 13:45:25 DESKTOP-4V4OMAL Browser-1[8092] INFO This time we are Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/32.0.1664.3 Safari/537.36
2021-11-20 13:45:25 DESKTOP-4V4OMAL Browser-1[8092] INFO Waiting 10s for page to load and js to run.
'C:\Program' is not recognized as an internal or external command,
operable program or batch file.
```

**Comment by 2e0byo** [comment]
Yeah, that's windows being predictably stupid.  Note how it *tried* to run "C:\Program".  Only windows is stupid enough to have a *space* in a system dirname.  The crude solution is to do the escaping yourself, with something like:

```python
r"C:\Program\ Files\tor.exe" # and the others
```
A better solution is to handle the escaping automatically.  I've just pushed up a commit to do this using `shlex.quote()` but I don't have windows on hand to test. Can you test it?

**Comment by chrishat34** [comment]
ok, not sure that has helped.....

```python

(venv) C:\Development from 16-11-2021\YADC>python -m yadc.main
2021-11-20 18:52:12 DESKTOP-4V4OMAL Browser-1[7688] INFO Starting search loop
2021-11-20 18:52:12 DESKTOP-4V4OMAL Browser-1[7688] INFO Starting Tor
2021-11-20 18:52:12 DESKTOP-4V4OMAL Browser-1[7688] ERROR [WinError 2] The system cannot find the file specified
Traceback (most recent call last):
  File "C:\Development from 16-11-2021\YADC\yadc\scraper.py", line 247, in __call__
    with self._browser as browser:
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 262, in __enter__
    self.launch_chrome()
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 348, in launch_chrome
    self.start_tor()
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 343, in start_tor
    proc = Popen(cmd, stdout=PIPE, encoding="utf8")
  File "C:\Development from 16-11-2021\python399\lib\subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Development from 16-11-2021\python399\lib\subprocess.py", line 1420, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified
Traceback (most recent call last):
  File "C:\Development from 16-11-2021\YADC\yadc\scraper.py", line 247, in __call__
    with self._browser as browser:
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 262, in __enter__
    self.launch_chrome()
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 348, in launch_chrome
    self.start_tor()
  File "C:\Development from 16-11-2021\YADC\yadc\browser.py", line 343, in start_tor
    proc = Popen(cmd, stdout=PIPE, encoding="utf8")
  File "C:\Development from 16-11-2021\python399\lib\subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Development from 16-11-2021\python399\lib\subprocess.py", line 1420, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Development from 16-11-2021\python399\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Development from 16-11-2021\python399\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Development from 16-11-2021\YADC\yadc\main.py", line 42, in <module>
    s()
  File "C:\Development from 16-11-2021\YADC\yadc\scraper.py", line 255, in __call__
    randsleep(30) # try not to be too predictable
  File "C:\Development from 16-11-2021\YADC\yadc\humanlike.py", line 27, in randsleep
    sleep(duration)
KeyboardInterrupt
^C
```

**Comment by 2e0byo** [comment]
Yes, your path is wrong (in the form python eventually sees, anyhow).  Add a print statement in browser.py right before the subprocess line (the error traceback will tell you the line number).  What command is it *trying* to run.  What should it be?  If you fire up a repl (just type `python` at the shell) you can try it by hand (`from subprocess import Popen` and then `Popen(cmd, shell=True)` where cmd is whatever it printed out earlier).

Sorry not to be more help: thus is ultimately a problem with the particular setup your end (unless I made a silly typo in the last commit: I'll check tomorrow).  If all else fails, copy chrome/chromedriver/tor into the YADC folder and use relative paths (I.e. `./chrome.exe`).  Unless your shell is being silly (always possible on Windows) that should work around any path issues.  But absolute paths are definitely possible.

As an aside, wouldn't it be nice if Windows told you *which* file it couldn't find...

**Comment by chrishat34** [comment]
Hi, I really appreciate all the help you have given, thank you, but im sorry i still cannot get it running on windows.
I'll keep trying.

**Comment by 2e0byo** [comment]
I suppose I'll have to spin up a windows VM and get it running tomorrow.  Won't be till the evening.  I'll update when I do.

**Comment by chrishat34** [comment]
That would be brilliant, thank you.

**Comment by 2e0byo** [comment]
Well windows is a complete nightmare, isn't it.

With the latest changes in this repo I am able to run this under windows.  My env was:

- python 3.9
- windows 10
- tor downloaded via `scripts/download_tor.py` (you will need to `pip install requests` first)
- chromedriver in the same dir
- chrome installed

Setup like this I was able to run the second two tests in `tests/test_browser` successfully, which is equivalent to running your `main.py` assuming everything is set up properly.

Changes are:

- drop use of `shutil.quote` since it apparently doesn't work on windows.
- use `Path()` objects when we dont use `shell=True`, and use manual quoting with *double* quotes otherwise
- get a new port if required for chromium (it got blocked on my end)

Thus it *is* possible to get this running on windows.  Note that the browser tests fail in github actions at the moment because I can't figure out how to launch chrome without a graphical server.  They run locally (although the first wil fail except Linux with `google-chrome-stable` installed).

**Comment by chrishat34** [comment]
That's excellent thank you, i have managed to get it running and looking like its checking for earlier tests. There are a few error messages that crop up still but i will test some more to try and resolve those before i bother you again, thank you once again.

**Comment by 2e0byo** [comment]
Excellent, I'll close this.  I suppose I should add some more thorough advice for windows.

Note that there are quite a few expected exceptions: in general if it doesn't crash things are good.  You can have a look in your errors dir to see tracebacks and screenshots for every caught error.  Most should be things like 'too old a browser' or 'captcha solving failed' or 'you got blocked'.  These will just trigger a restart using a new tor connector and browser user agent, hopefully working around the problem.  Any errors actually in the *source* are bugs, so reporting them would be helpful.

Note that if I get a chance to get the tests running in ci (should be possible with headless chrome, I just couldn't get it going) I'll likely consider this project finished.  I'll still merge PRs though.

**Comment by chrishat34** [comment]
hi 2e0byo
I hope this is the right place to ask this.

In the main.py bit of code, i have a query on how to specify dates:

```python
driver = Driver(
    licence_number="YOUR_NUMBER",
    booking_ref="YOUR_REF",
    not_before=datetime(2021, 12, 10),
    not_after=datetime(2022, 3, 1),
    centres=[
        Centre(centre="Gateshead", before=datetime(2022, 1, 1)),
        Centre(centre="Durham"),
    ],
)
```
Lets say a test is booked for 15/01/2021
what would the datetime configuration be to look for dates after the booked test? say between 01/03/21 and 01/04/2021
Also im not sure what the before=datetime(2022, 1, 1) means in the centres list as its possibly defined earlier in the not_before... section?
Any advice would be much appreciated.

**Comment by 2e0byo** [comment]
@chrishat34 this belongs in a new issue: see the linked issue.
